[TASK] autoconfig WSC simple agent -> controller flow

tomereli opened the pull request:
The first stage in autoconfig flow - replacing slave join message with
autoconfig WSC simple M1 message by updating son slave thread
STATE_JOIN_MASTER to create an
AP_AUTOCONFIGURATION_WIFI_SIMPLE_CONFIGURATION_MESSAGE cmdu message and
add the Intel vendor specific TLV which includes all of the slave join
data, instead of using a vendor specific cmdu.
The controller, upon receiving this message, simply calls
handle_slave_join() after aligning the beerocks header with the M1
parameters.

**Agent changes summary**
- Send WSC CMDU with VS TLV instead of slave join in a VS CMDU

**Controller changes summary**
- subscribe to autoconfig simple message type
- add handle_cmdu_1905_autoconfiguration_WSC() which calls
handle_slave_join()

Issue #71

Commit:
agent: replace slave join with autoconfig WSC M1

First stage in autoconfig flow - replacing slave join message with
autoconfig WSC simple M1 message by updating son slave thread
STATE_JOIN_MASTER to create an
AP_AUTOCONFIGURATION_WIFI_SIMPLE_CONFIGURATION_MESSAGE cmdu message and
add the intel vendor specific TLV which includes all of the slave join
data, instead of using a vendor specific cmdu.
The controller, upon receiving of this message, simply calls
handle_slave_join() after aligning the beerocks header with the M1
parameters.

Agent changes summary:
- Send WSC CMDU with VS TLV instead of slave join in a VS CMDU

Controller changes summary:
- subscribe to autoconfig simple message type
- add handle_cmdu_1905_autoconfiguration_WSC() which calls
handle_slave_join()

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

Commit:
framework: tlvf: CmduMessageTx add add_vs_tlv()

Replace create method which accepts vendor spcific tlv OUI with
add_vs_tlv() API which allows callers to add vendor specific TLV
to a non vendor specific CMDU.

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

Commit:
common: tlvf:  add add_intel_data() API

Add helper API to add Intel vendor specific data (Action header
and operation), so it can be used for adding Intel vendor
specific TLV on none vendor specific CMDUs.

While at it, refactor create_vs_message() so it will use the
newly added add_intel_vs_data().

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

tomereli commented:
I'm opening this PR to review since it is working and we can actually merge it. This is the plan - keep things working and move forward. It is not perfect, so any review is highly appreciated.

tomereli commented:
First commit to review - https://github.com/prplfoundation/prplMesh/pull/66/commits/f0e5e5ed2a9feddab83811635eeee58266ed04cc

simonbaren commented on controller/src/beerocks/master/son_master_thread.cpp:
```diff
@@ -205,6 +206,10 @@ void master_thread::handle_cmdu_1905_1_message(Socket *sd, ieee1905_1::CmduMessa
         handle_cmdu_1905_autoconfiguration_search(sd, cmdu_rx);
         break;
     }
+    case ieee1905_1::eMessageType::AP_AUTOCONFIGURATION_WIFI_SIMPLE_CONFIGURATION_MESSAGE: {
```
Maybe we should keep the names as in the standard:
`AP_AUTOCONFIGURATION_WSC_MESSAGE`

tomereli replied:
Maybe, its even a shorter name so will get better look with clang-format, but not related to this PR (this was already merged and reviewed by @arnout )

simonbaren commented on framework/tlvf/AutoGenerated/include/tlvf/CmduMessageTx.h:
```diff
@@ -26,7 +26,7 @@ class CmduMessageTx : public CmduMessage {
 
 public:
```
Please commit the AutoGenerated files in different commit

tomereli replied:
In cases where there are not that many auto generated files, I find it easier to keep one commit for both. I think this is the case here - but if you and @arnout think otherwise, I'll split it.

arnout replied:
As you know, my POV is that AutoGenerated should not be included at all :-)

The only reason to make it a separate commit is to make review easier - but it doesn't make review that much easier, so I'm fine with it either way.

simonbaren commented on common/beerocks/tlvf/AutoGenerated/include/beerocks/tlvf/beerocks_message.h:
```diff
@@ -70,51 +70,71 @@ class message_com {
 
     template <class T>
-    static std::shared_ptr<T> create_vs_message(ieee1905_1::CmduMessageTx &cmdu_tx, uint16_t id = 0)
+    static std::shared_ptr<T> add_intel_vs_data(ieee1905_1::CmduMessageTx &cmdu_tx, uint16_t id = 0)
```
Also here: commit AutoGenerated in different  commit

tomereli replied:
Same answer - lets see what @arnout has to say

simonbaren commented on common/beerocks/tlvf/src/include/beerocks/tlvf/beerocks_message.h:
```diff
@@ -67,51 +67,71 @@ class message_com {
     }
 
+    template <class T>
+    static std::shared_ptr<T> create_vs_message(ieee1905_1::CmduMessageTx &cmdu_tx, uint16_t id = 0)
```
Suggestion:
`create_vs_message` should only create the VS CMDU,
After that we should call for `add_intel_vs_data<T>()` separately

tomereli replied:
Agree, but its a huge change which I didn't want to make now just to add support for other vendors to use `create_vs_message` - not our focus.

simonbaren commented on agent/src/beerocks/slave/son_slave_thread.cpp:
```diff
@@ -3789,7 +3791,7 @@ bool slave_thread::slave_fsm(bool &call_slave_select)
     }
     case STATE_JOIN_MASTER: {
-
+        LOG(ERROR) << "LIOR JOIN MASTER";
```
I'm sure that Lior didn't actually joined the master :)

simonbaren commented on agent/src/beerocks/slave/son_slave_thread.cpp:
```diff
@@ -3798,12 +3800,44 @@ bool slave_thread::slave_fsm(bool &call_slave_select)
+        if (tlvAp == nullptr) {
+            LOG(ERROR) << "Error creating TLV_AP_RADIO_BASIC_CAPABILITIES";
+        }
+        for (int i=0; i<6; i++) {
```
what is 6?
I guess it's the MAC_ADDR_LEN?

simonbaren commented on agent/src/beerocks/slave/son_slave_thread.cpp:
```diff
@@ -3798,12 +3800,44 @@ bool slave_thread::slave_fsm(bool &call_slave_select)
+            LOG(ERROR) << "Error creating TLV_AP_RADIO_BASIC_CAPABILITIES";
+        }
+        for (int i=0; i<6; i++) {
+            tlvAp->radio_uid().mac[i] = network_utils::mac_from_string(config.radio_identifier).oct[i]; // TODO: put mac_from_string in variable
```
Instead of this loop you should use:
`std::copy_n(network_utils::mac_from_string(config.radio_identifier).oct, MAC_ADDR_LEN,  tlvAp->radio_uid().mac)`

BTW: the standard doesn't define that the RUID==mac. Yes. it's the same size as MAC but not exclusive to that.
Please review the YAML file of this TLV to make sure it is up with the standard (**section 17.2.7  AP Radio Basic Capabilities TLV Format**)

simonbaren commented on agent/src/beerocks/slave/son_slave_thread.cpp:
```diff
@@ -3798,12 +3800,44 @@ bool slave_thread::slave_fsm(bool &call_slave_select)
+        for (int i=0; i<6; i++) {
+            tlvAp->radio_uid().mac[i] = network_utils::mac_from_string(config.radio_identifier).oct[i]; // TODO: put mac_from_string in variable
+        }
+        tlvAp->maximum_number_of_bsss_supported() = beerocks::IFACE_TOTAL_VAPS; //TODO: make sure this is the write value
```
` //TODO: make sure this is the **right** value`

simonbaren commented on agent/src/beerocks/slave/son_slave_thread.cpp:
```diff
@@ -3798,12 +3800,44 @@ bool slave_thread::slave_fsm(bool &call_slave_select)
+            operationClassesInfo->maximum_transmit_power_dbm() = 24; //TODO: take value from stats
+            for (int j = 0; j< 2; j++) {
+                if (operationClassesInfo->alloc_statically_non_operable_channels_list()) {
+                    std::get<1>(operationClassesInfo->statically_non_operable_channels_list(i)) = i+j+10; //TODO blah blah
```
Great TODO :)

simonbaren commented on controller/src/beerocks/master/son_master_thread.cpp:
```diff
@@ -208,6 +208,7 @@ void master_thread::handle_cmdu_1905_1_message(Socket *sd, ieee1905_1::CmduMessa
     }
     case ieee1905_1::eMessageType::AP_AUTOCONFIGURATION_WIFI_SIMPLE_CONFIGURATION_MESSAGE: {
         LOG(DEBUG) << "Received AP_AUTOCONFIGURATION_WIFI_SIMPLE_CONFIGURATION_MESSAGE";
+        handle_cmdu_1905_autoconfiguration_WSC(sd, cmdu_rx);
```
If you changed to `bool handle_cmdu_1905_autoconfiguration_WSC` shouldn't you check the return value?

tomereli replied:
This is for future validity checking which is irrelevant now for controller certificaiton.

arnout replied:
I'm with Semyon here.

Most of the other handle() functions return void because the caller doesn't actually handle errors. Anyway, if the handling "fails", it is not an error state that the caller could handle. You'll just not get an updated flow state.

If you do return something, it looks as if the caller is expected to handle something. It just makes the code harder to understand that way.

Also, at some point it would be good to enable the 'unused return value' warning. If a function returns something, it should be used.

simonbaren commented on controller/src/beerocks/master/son_master_thread.cpp:
```diff
@@ -356,25 +357,33 @@ void master_thread::handle_cmdu_1905_autoconfiguration_search(Socket *sd,
     }
+    
+    // TODO autoconfig remove slave join
+    std::memcpy(beerocks_header->radio_mac().oct, tlvAp->radio_uid().mac, 6);
```
`std::copy_n`
6 is magic. use MAC_ADDR_LEN

simonbaren commented on controller/src/beerocks/master/son_master_thread.cpp:
```diff
@@ -356,25 +357,33 @@ void master_thread::handle_cmdu_1905_autoconfiguration_search(Socket *sd,
+    // TODO autoconfig remove slave join
+    std::memcpy(beerocks_header->radio_mac().oct, tlvAp->radio_uid().mac, 6);
+    LOG(INFO) << "Handle slave join, radio identifier " <<
+        network_utils::mac_to_string(beerocks_header->radio_mac());
```
there is a utility:
`static std::string mac_to_string(const uint8_t *mac_address);`

tomereli replied:
not sure what you mean - I am using this utility

Commit:
common: bwl: dummy: get_radio_mac() support

Implement get_radio_mac() in bwl dummy which reads the linux interface
MAC address.

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

tomereli commented:
operational - ready for review and merge

arnout reviewed (commented):
Commit message of the first patch: typo WSA -> WSC.

arnout commented:
Patch 1 (controller: initial handling for autoconfig WSC) and 6 (controller: initial handling for autoconfig WSC) should be squashed.

arnout commented on controller/src/beerocks/master/son_master_thread.cpp:
```diff
@@ -351,6 +356,27 @@ void master_thread::handle_cmdu_1905_autoconfiguration_search(Socket *sd,
+    if (!beerocks_header) {
+        return; //not an intel radio
+    }
+
```
Don't you also need to check `beerocks_header->direction() == beerocks::BEEROCKS_DIRECTION_AGENT`?

tomereli replied:
Basically yes - since this message can originate from both the controller and the agent apparently:

> If the message is sent by the Multi-AP Agent:
>  - One AP Radio Basic Capabilities TLV (see section 17.2.7). 
>  - One WSC TLV (containing M1).
> If the message is sent by the Multi-AP Controller:
>   - One AP Radio Identifier TLV (see section 17.2.3). 
>   - One or more WSC TLV (containing M2). 

So it is better for the future (when we will get rid of the vendor specific parts which are redundant) to do this properly - I will fix the code to support this.

arnout commented on controller/src/beerocks/master/son_master_thread.cpp:
```diff
@@ -351,6 +356,27 @@ void master_thread::handle_cmdu_1905_autoconfiguration_search(Socket *sd,
+    if (beerocks_header->action_op() ==
+        beerocks_message::ACTION_CONTROL_SLAVE_JOINED_NOTIFICATION) {
+        handle_slave_join(sd, beerocks_header,
+                                 cmdu_rx); // TODO to be removed once autoconfig is completed
```
Actually, it makes sense that it stays. What should be removed is the "not an intel radio" exit.

arnout commented on common/beerocks/tlvf/src/include/beerocks/tlvf/beerocks_message.h:
```diff
@@ -67,51 +67,73 @@ class message_com {
+    template <class T>
+    static std::shared_ptr<T> create_vs_message(ieee1905_1::CmduMessageTx &cmdu_tx, uint16_t id = 0)
+    {
+        auto cmduhdr = cmdu_tx.create(id, ieee1905_1::eMessageType::VENDOR_SPECIFIC_MESSAGE);
```
This change should also have been done in the previous patch that removed the `create(OUI)` function. Otherwise, reverting or bisecting this patch would break. So, the commit that adds add_vs_tlv should also use it in this function, and this commit should only move out the populating of the TLV to the new add_intel_vs_data function.

Ideally, Travis should test each individual commit, which would have exposed this problem.

:ok:  BTW, you're changing the name of the variable for no good reason, and you're also switching the condition from `if (cmduhdr)` to `if (!cmduhdr)`. Such changes make the diff bigger and review more difficult. BTW2, I'm experimenting here by adding the :ok: marker to indicate that this is a comment you don't need to take action on :-)

arnout commented on agent/src/beerocks/slave/son_slave_thread.cpp:
```diff
@@ -475,9 +478,9 @@ bool slave_thread::handle_cmdu_control_message(
-                    hostap_params.iface_mac.oct)) {
+    // to me or not to me, this is the question...
+    if (network_utils::mac_to_string(beerocks_header->radio_mac()) !=
+        network_utils::mac_to_string(hostap_params.iface_mac)) {
```
Why is it better to do mac_to_string than to use binary comparison? The commit message doesn't enlighten me.

It would be better if sMacAddr would have an equality operator that does this binary comparison.

While I'm at it: why does sMacAddr have the channel and rssi members?

tomereli replied:
Agree, I didn't want to do that since it is not related to this change but since you brought it up i'll do the change, in a separate commit.
I have no idea why sMacAddr have channel and rssi - probably someone who was too lazy to define a wrapper struct - I will fix this as well.

arnout commented on agent/src/beerocks/slave/son_slave_thread.cpp:
```diff
@@ -3798,12 +3801,64 @@ bool slave_thread::slave_fsm(bool &call_slave_select)
+            }
+        }
+
+        //TODO: add WSC tlv with M1
```
This comment seems spurious - that's exactly what you do here, no?

arnout commented on agent/src/beerocks/slave/son_slave_thread.cpp:
```diff
@@ -3798,12 +3801,64 @@ bool slave_thread::slave_fsm(bool &call_slave_select)
+        string_utils::copy_string(tlvWscM1->M1Frame().model_name_attr.data, "Ubuntu", 7);
+        string_utils::copy_string(tlvWscM1->M1Frame().model_number_attr.data, "18.04", 6);
+        string_utils::copy_string(tlvWscM1->M1Frame().device_name_attr.data, "prplMesh", 9);
+
```
M1 should also have values for:

* uuid_e_attr
* enrolee_nonce_attr -> to be added by encryption, but the TODO should be here already
* public_key_attr -> same
* authentication_type_flags_attr
* encryption_type_flags_attr
* serial_number_attr
* primary_device_type_attr -> We could actually add default values to the yaml
* rf_bands_attr -> This is essential and must correspond to the freqband in the search message
* vendor_extensions_attr -> this could also be done in the yaml file

All of the above could be added in separate commits/PRs, but it would be good to have the TODOs here. Also, it would be good to add an explicit comment that all the other attributes are set with correct defaults from the yaml defintion.

Finally, it would be better to move this M1 population to a separate function, similar to add_intel_vs_data.

arnout commented on agent/src/beerocks/slave/son_slave_thread.cpp:
```diff
@@ -3875,6 +3930,8 @@ bool slave_thread::slave_fsm(bool &call_slave_select)
         // Channel Selection Params
         notification->cs_params() = hostap_cs_params;
+
+        tlvVendorSpecific->length() += notification->getLen();
```
It would make more sense to do this inside add_intel_vs_data

tomereli replied:
I wanted to do that but if there are variable length list in the TLV then we have to set it here.
We could set the initial size inside `add_intel_vs_data` and the caller will just need to update the length with the allocations.

arnout replied:
Ow, my bad, I wasn't paying attention at all apparently. Obviously indeed this can only be done here.

arnout commented on common/beerocks/bwl/dummy/base_wlan_hal_dummy.cpp:
```diff
@@ -113,7 +113,15 @@ bool base_wlan_hal_dummy::refresh_vaps_info(int id) { return true; }
+    }
+    return mac;
+}
+//std::string base_wlan_hal_dummy::get_radio_mac() { return std::string("DE:AD:BE:EF:DE:AD"); }
```
Remove commented-out code.

arnout commented on controller/src/beerocks/master/son_master_thread.cpp:
```diff
@@ -351,6 +358,41 @@ void master_thread::handle_cmdu_1905_autoconfiguration_search(Socket *sd,
+        return false;
+    }
+
+    auto tlvWscM1 = cmdu_rx.addClass<ieee1905_1::tlvWscM1>();
```
Here we're again hitting a limit of tlvf...

The message you receive may be an M1 or an M2. The only way to know is by looking at the message_type_attr. addClass however only looks at the type and length, not at the contents.

The best you can do for now I think is to add a condition a check on message_type_attr below. In practice, you'll never hit it because the M2 has a different size so you'll get an "Error creating tlvWscM1" instead.

tomereli replied:
How can the controller receive M2?

arnout replied:
If there is an agent on the same device, the destination AL-MAC is the device itself, so the ieee1905_transport will publish the CMDU back on the local bus. Since the CMDU is of type WSC (not WSC M2), the controller is subscribed to it so it will receive its own CMDU.

arnout replied:
Note that this is also already handled by looking at the Radio Info / Radio Identifier TLV, so it's OK for now.

arnout reviewed (changes requested):
Commit "autoconfig simple message type subscribe" commit message is not OK. E.g.:
"controller: subscribe to WSC message type".
However, it doesn't make much sense to do this in a separate commit.

Actually, the whole commit series is not properly split up. It makes it almost impossible to review this series. The best that I can do is to review the end result, but treating a +251-110 change as a single blob is also pretty heavy.

tomereli commented:
Since this PR is not well organized into commits, I will add fixes in new commits so it will be reviewable, then once it is accepted I will squash it all, then split it up nicely. @arnout won't be here to see it happen but I'll do my best to split it logically :)

arnout commented:
> Since this PR is not well organized into commits, I will add fixes in new commits so it will be reviewable, then once it is accepted I will squash it all, then split it up nicely. @arnout won't be here to see it happen but I'll do my best to split it logically :)

As I mentioned in private: I don't think it's worth spending much effort to split into commits. These are the reasons why splitting up commits is important:

* It is more inviting for review (a 1000-line patch looks impossible to review, a 10-line patch is quickly done).
* It makes it possible to merge the first part already and leave the last part with change requests to rework.
* It makes it possible to backport a fix to a release branch.
* It makes it possible to revert only the bad part of a bad patch.
* It makes it easier to bisect to a problem and to identify the real cause.
* It allows you to test changes incrementally instead of a big bang change where debugging breakage is complicated.
* It makes it easier to understand why a particular change was done.
* It makes it possible to document the code with commit logs instead of comments.
* It makes it easier to look at the history and understand what has been done recently.

For a feature commit like this one, most of these reasons don't apply. It is mostly the review and documentation aspects that are relevant. Since the review aspect no longer applies, I don't think splitting up is very useful. Perhaps split up in 3 patches:

1. create() -> add_vs_tlv()
2. split off add_intel_vs_data() from create_vs_message()
3. the rest

But even that does not bring so much added value. And even this split is quite some work - you can't simply use `git commit -p` because e.g. the code calling add_vs_tlv() looks completely different than in the original. And what we definitely don't want is to split it into commits where the intermediate commits are broken - that goes against almost all of the reasons to split up commits.

Note that in general, I am absolutely in favour of creating very fine grained commits. For sure, while developing, people should already try to commit things in a logical way, for many of the reasons mentioned above. However, in this phase of the project, where we're doing big feature development, many of the reasons mentioned above don't apply anyway. Therefore, once you've gone and messed up the commit sequence, I think cleaning that up is just overhead with little added value.

arnout commented on controller/src/beerocks/master/son_master_thread.cpp:
```diff
@@ -332,38 +331,41 @@ void master_thread::handle_cmdu_1905_autoconfiguration_search(Socket *sd,
     auto tlvAp = cmdu_rx.addClass<wfa_map::tlvApRadioBasicCapabilities>();
     if (tlvAp == nullptr) {
-        LOG(ERROR) << "Failed to get APRadioBasicCapabilities";
+        // Check if this a message we sent to the agent which was just looped back
```
This needs a better comment... You need to explain why it would be looped back in the first place, and why you can use the radio identifier to detect loopback (which you can't, actually, see below). Something like:

```
// Check if this is a M2 message that we sent to the agent, which was just looped back.
// The M1 and M2 messages are both of CMDU type AP_Autoconfiguration_WSC. Thus,
// when we send the M2 to the local agent, it will be published back on the local bus because
// the destination is our AL-MAC, and the controller does listen to this CMDU.
// Ideally, we should use the message type attribute from the WSC payload to distinguish.
// Unfortunately, that is a bit complicated with the current tlv parser. However, there is another
// way to distinguish them: the M1 message has the AP_Radio_Basic_Capabilities TLV,
// while the M2 has the AP_Radio_Identifier TLV.
// If this is a looped back M2 CMDU, we can treat is as handled successfully.
```

Unfortunately this distinction is not entirely correct, because the M1 may come from a non-MultiAP 1905.1 device. In that case, there is no AP Radio Basic Capabilities in M1 and no AP Radio Identifier in M2. But that is not something we have to handle now. :ok:

arnout commented on controller/src/beerocks/master/son_master_thread.cpp:
```diff
@@ -374,10 +376,12 @@ bool master_thread::handle_cmdu_1905_autoconfiguration_WSC(Socket *sd,
     }
 
     //TODO autoconfig process the rest of the class
+    //TODO autoconfig add support for none intel agents
```
The correct spelling is non-Intel, not "none".

tomereli replied:
will fix in the squash

Commit:
common: refactor net_struct sMacAddr

sMacAddr was used for MAC Address AND for scan results. This is just wrong,
so change sMacAddr to be just a MAC address
struct, and add a new struct for the scan results.
While we're at it - add operator== and operator!= implementations
to all of the structs defined in net_struct.h.

Changes caused changes in the common, agent and controller,
and are all part of this commit.

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

arnout commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:
```diff
@@ -1564,7 +1564,7 @@ bool main_thread::handle_slave_backhaul_message(std::shared_ptr<SSlaveSockets> s
                 break;
             }
-            response->params().mac        = request->params().mac;
+            response->params().result.mac        = request->params().mac;
```
I don't think you've run clang-format here :-)

tomereli replied:
will run clang-format on each commit before merging - keeping this open till I do

tomereli replied:
ok it seems that we are not aligned to clang-format anymore - so I'll merge this as is now and re-run clang-format in a separate PR which also enables clang-format check in travis CI.
See #97

arnout commented on common/beerocks/bcl/include/beerocks/bcl/network/net_struct.h:
```diff
@@ -23,28 +24,44 @@ enum eNetworkStructsConsts {
+    bool operator!=(sMacAddr const & rhs) const { return !(rhs == *this); }
+} __attribute__((packed)) sMacAddr;
+
+typedef struct sScanResult {
```
There is also a struct bwl::SScanResult with essentially the same contents, and in backhaul_manager_thread one is copied to the other. Wouldn't it make more sense to use just one struct?

arnout commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:
```diff
@@ -2178,7 +2178,9 @@ void main_thread::get_scan_measurement()
-            auto it = scan_measurement_list.find(bssid);
+            auto it = std::find_if(scan_measurement_list.begin(), scan_measurement_list.end(),
+                        [bssid](const std::pair<std::string, net::sScanResult> & t) -> bool
+                        { return t.second.mac == network_utils::mac_from_string(bssid); });
```
I must be missing something major, but why doesn't find() work any more? The key is still the bssid, right?

If this is needed, there is no point of making it a map any more - the bssid key is not used anywhere else.

arnout commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:
```diff
@@ -2190,9 +2192,9 @@ void main_thread::get_scan_measurement()
+                sScanResult scan_measurement;
 
-                std::copy_n(scan_result.bssid, sizeof(sMacAddr::oct), scan_measurement.oct);
+                std::copy_n(scan_result.bssid, MAC_ADDR_LEN, scan_measurement.mac.oct);
```
Personally, I prefer the `sizeof()`.

arnout reviewed (approved):
Some small changes requested still but OK. I tested myself and I also see the M1.

Please don't forget to remove the commented-out base_wlan_hal_dummy::get_radio_mac as well.

Commit:
common: bwl: use beerocks::net::sScanResult

Now that we have defined a sScanResult struct, use it instead of
redefining it in the BWL.

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>
